src/mesh/boundary_nonuniform.cxx

```diff
@@ -0,0 +1,1285 @@
+#include "boundary_nonuniform.hxx"
+
+#include <bout/constants.hxx>
+
+#include <boundary_standard.hxx>
+#include <boutexception.hxx>
+#include <derivs.hxx>
+#include <fft.hxx>
+#include <globals.hxx>
+#include <invert_laplace.hxx>
+#include <msg_stack.hxx>
+#include <output.hxx>
+#include <utils.hxx>
+
+void BoundaryDirichletNonUniform_O2::apply(Field3D& f, BoutReal t) {
+  bndry->first();
+
+  // Decide which generator to use
+  std::shared_ptr<FieldGenerator> fg = gen;
+  if (!fg)
+    fg = f.getBndryGenerator(bndry->location);
+
+  BoutReal val = 0.0;
+
+  Mesh* mesh = f.getMesh();
+  CELL_LOC loc = f.getLocation();
+
+  BoutReal vals[mesh->LocalNz];
+  int bx = bndry->bx;
+  int by = bndry->by;
+  int stag = 0;
+  if (loc == CELL_XLOW) {
+    if (bx == 0) {
+      bx = -1;
+    } else if (bx < 0) {
+      stag = -1;
+    } else {
+      stag = 1;
+    }
+  }
+  if (loc == CELL_YLOW) {
+    if (by == 0) {
+      by = -1;
+    } else if (by < 0) {
+      stag = -1;
+    } else {
+      stag = 1;
+    }
+  }
+  int istart = (stag == -1) ? -1 : 0;
+
+  for (; !bndry->isDone(); bndry->next1d()) {
+    if (fg) {
+      for (int zk = 0; zk < mesh->LocalNz; zk++) {
+        // Calculate the X and Y normalised values half-way between the guard cell and
+        // grid cell
+        BoutReal xnorm = 0.5
+                         * (mesh->GlobalX(bndry->x)          // In the guard cell
+                            + mesh->GlobalX(bndry->x - bx)); // the grid cell
+
+        BoutReal ynorm = 0.5
+                         * (mesh->GlobalY(bndry->y)          // In the guard cell
+                            + mesh->GlobalY(bndry->y - by)); // the grid cell
+
+        vals[zk] = fg->generate(xnorm, TWOPI * ynorm, TWOPI * zk / (mesh->LocalNz), t);
+      }
+    }
+    BoutReal fac0;
+    BoutReal x0;
+    BoutReal fac1;
+    BoutReal x1;
+
+    const Field2D& dy =
+        bndry->by != 0 ? mesh->getCoordinates()->dy : mesh->getCoordinates()->dx;
+    Indices i1{bndry->x - 1 * bndry->bx, bndry->y - 1 * bndry->by, 0};
+    BoutReal t;
+    if (stag == 0) {
+      x0 = 0;
+      BoutReal st = 0;
+      t = dy[i1];
+      x1 = st + t / 2;
+      st += t;
+    } else {
+      x0 = 0; // dy(bndry->x, bndry->y) / 2;
+      x1 = x0 + dy[i1];
+    }
```

Comment:
This code is also repeated a lot -- can it be pulled out?

When we eventually get the boundary stuff into a framework like the derivatives, this bit will be something like `populateStencil`. Can we do something like that now? Put `x0`, `x1` and `fac0`, `fac1` into separate stencils/structs?

Reply:
There are slight differences, depending on what BC is used. Thus I kept it for now.
The more trivial for loop has been removed.